src/Rings/oscar_singular.jl

```diff
@@ -0,0 +1,458 @@
+##############################################################################
+#
+# Conversion to and from Singular: in particular, some Rings are
+# special as they exist natively in Singular and thus should be used
+#
+##############################################################################
+#
+# Needs convert(Target(Ring), elem)
+# Ring(s.th.)
+#
+# Singular's polynomial rings are not recursive:
+# 1. singular_poly_ring(R::Ring) tries to create a Singular.PolyRing (with
+#    elements of type Singular.spoly) isomorphic to R
+# 2. singular_coeff_ring(R::Ring) tries to create a ring isomorphic to R that is
+#    acceptable to Singular.jl as 'coefficients'
+#
+# a real native Singular polynomial ring with Singular's native QQ as coefficients:
+#  singular_poly_ring(QQ[t]) => Singular.PolyRing{Singular.n_Q}
+#
+# Singular's native Fp(5):
+#  singular_coeff_ring(GF(5)) => Singular.N_ZpField
+#
+# Singular wrapper of the Oscar type QQPolyRingElem:
+#  singular_coeff_ring(QQ[t]) => Singular.N_Ring{QQPolyRingElem}
+#
+# even more wrappings of the immutable Oscar type FpFieldElem:
+#  singular_coeff_ring(GF(ZZRingElem(5))) => Singular.N_Field{Singular.FieldElemWrapper{FpField, FpFieldElem}}
+
+"""
+    iso_oscar_singular_coeff_ring(R::Ring) -> Map
+
+Return a ring isomorphism of `R` onto a native Singular ring
+"""
+iso_oscar_singular_coeff_ring
+
+"""
+    iso_oscar_singular_poly_ring(R::Ring, ...; kw...) -> Map
+
+Given a polynomial ring `R[x_1,...x_n]` return a ring isomorphism onto a native
+Singular ring `S[x_1,...,x_n]` with `R` isomorphic to `S`.
+"""
+iso_oscar_singular_poly_ring
+
+abstract type OscarSingularCoefficientRingMap{S, T} <: Map{S, T, Any, Any} end
```

Comment:
Very small nitpick: I was briefly confused because `S` here is not the same as `f.S` later (i.e. `S` as two meanings). How about following the naming used in AA here, which hint sat the use, "Domain type" and "Codomain type"?
```suggestion
abstract type OscarSingularCoefficientRingMap{D, C} <: Map{D, C, Any, Any} end
```